Placeholder PR: Blog for Volume Group Snapshot (#39415)
kubernetes/website: content/en/blog/_posts/2023-04-11-volume-group-snapshot-alpha.md

---
layout: blog

> title: "Kubernetes 1.27: Volume Group Snapshot Moves to Alpha"

Reviewer: Do we have some of this details on the k8s.io/docs website?

Author: The doc will be at the CSI driver doc website: https://kubernetes-csi.github.io/docs/

Author: We don't have it yet. We'll have the doc after external-snapshotter and other CSI sidecars are released which will happen after K8s 1.27 release.

Author: Added a WIP doc PR here: https://github.com/kubernetes-csi/docs/pull/557

date: 2023-04-11T10:00:00-08:00
slug: kubernetes-1-27-volume-group-snapshot-alpha
---

**Author:** Xing Yang (VMware)

> Volume Group Snapshot is introduced as an alpha feature in Kubernetes v1.27. The Volume Group Snapshot feature introduces a Kubernetes API that allows users to take a crash consistent snapshot of multiple volumes together. It uses a label selector to group multiple persistent volume claims for snapshotting. This new feature is only supported for CSI Volume Drivers.

Reviewer: Please **manually wrap** the long lines for ease of review feedbacks and for ease of change tracking, especially for downstream localization teams.

e.g.
```suggestion
Volume group snapshot is introduced as an Alpha feature in Kubernetes v1.27.
This feature introduces a Kubernetes API that allows users to take a crash consistent
snapshot for multiple volumes together. It uses a label selector to group multiple
PersistentVolumeClaims for snapshotting.
This new feature is only supported for CSI volume drivers.
```

Reviewer 2: Please actually wrap the text.

Author: Done

It is true that we can quiesce the application first, take an individual snapshot from each volume that is part of the application one after the other, and then unquiesce the application after all the individual snapshots are taken. This way we will get application consistent snapshots. However, application quiesce is time consuming. Sometimes it may not be possible to quiesce an application. Taking individual snapshots one after another may also take longer time compared to taking a consistent group snapshot. Some users may not want to do application quiesce very frequently for these reasons. For example, a user may want to run weekly backups with application quiesce and nightly backups without application quiesce but with consistent group support which provides crash consistency across all volumes in the group.

## Kubernetes Volume Group Snapshots API

> Kubernetes Volume Group Snapshots introduce [three new API objects](https://github.com/kubernetes-csi/external-snapshotter/blob/master/client/apis/volumegroupsnapshot/v1alpha1/types.go) for managing snapshots:

Reviewer 2: If we can, link to the overview page and not the source code.

Author: The doc page is not ready yet. We'll have the doc after releasing the external-snapshotter and other CSI sidecars which we usually do after the K8s release.

Author: Added a WIP doc PR here: https://github.com/kubernetes-csi/docs/pull/557

> VolumeGroupSnapshot
> * Created by a Kubernetes user to request creation of a volume group snapshot for multiple volumes. It contains information about the volume group snapshot operation such as the timestamp when the volume group snapshot was taken and whether it is ready to use.
> The creation and deletion of this object represents a user’s desire to create or delete a cluster resource (a group snapshot).
> 
> VolumeGroupSnapshotContent
> * Created by the Snapshot Controller for a dynamically created VolumeGroupSnapshot. It contains information about the volume group snapshot including the volume group snapshot ID.
> This object represents a provisioned resource on the cluster (a group snapshot).
> The VolumeGroupSnapshotContent object binds to the VolumeGroupSnapshot for which it was created with a one-to-one mapping.
> 
> VolumeGroupSnapshotClass
> * Created by cluster administrators to describe how volume group snapshots should be created. including the driver information, the deletion policy, etc.

Reviewer 2: ```suggestion
`VolumeGroupSnapshot`
: Created by a Kubernetes user (or perhaps by your own automation) to request creation of a volume group snapshot for multiple volumes. It contains information about the volume group snapshot operation such as the timestamp when the volume group snapshot was taken and whether it is ready to use.
The creation and deletion of this object represents a desire to create or delete a cluster resource (a group snapshot).

`VolumeGroupSnapshotContent`
: Created by the snapshot controller for a dynamically created VolumeGroupSnapshot. It contains information about the volume group snapshot including the volume group snapshot ID.
This object represents a provisioned resource on the cluster (a group snapshot).
The VolumeGroupSnapshotContent object binds to the VolumeGroupSnapshot for which it was created with a one-to-one mapping.

`VolumeGroupSnapshotClass`
: Created by cluster administrators to describe how volume group snapshots should be created. including the driver information, the deletion policy, etc.
```

Author: Done.

The Volume Group Snapshot objects are defined as CustomResourceDefinitions (CRDs). These CRDs must be installed in a Kubernetes cluster for a CSI Driver to support volume group snapshots.

## How do I use Kubernetes Volume Group Snapshots

> Volume Group Snapshot feature is implemented in the external-snapshotter repo. It contains the following components:

Reviewer 2: ```suggestion
Volume Group Snapshot feature is implemented in the
[external-snapshotter](https://github.com/kubernetes-csi/external-snapshotter) repository. Implementing volume
group snapshots meant adding or changing several components:
```

Author: Done

* Kubernetes Volume Group Snapshot CRDs
* Volume group snapshot controller logic is added to the common snapshot controller.
* Volume group snapshot validation webhook logic is added to the common snapshot validation webhook.
* Logic to make CSI calls is added to CSI Snapshotter sidecar controller.

The volume snapshot controller, CRDs, and validation webhook are deployed once per cluster, while the sidecar is bundled with each CSI driver.

Therefore, it makes sense to deploy the volume snapshot controller, CRDs, and validation webhook as a cluster addon. It is strongly recommended that Kubernetes distributors bundle and deploy the volume snapshot controller, CRDs, and validation webhook as part of their Kubernetes cluster management process (independent of any CSI Driver).

> ### Creating a new group snapshot with Kubernetes

Reviewer: please always leave an empty line after a section or subsection.

Author: Done

Once a VolumeGroupSnapshotClass object is defined and you have volumes you want to snapshot together, you may create a new group snapshot by creating a VolumeGroupSnapshot object.

The source of the group snapshot specifies whether the underlying group snapshot should be dynamically created or if a pre-existing VolumeGroupSnapshotContent should be used. One of the following members in the source must be set.
* Selector - Selector is a label query over persistent volume claims that are to be grouped together for snapshotting. This labelSelector will be used to match the label added to a PVC.
* VolumeGroupSnapshotContentName - specifies the name of a pre-existing VolumeGroupSnapshotContent object representing an existing volume group snapshot.

For dynamic provisioning, a selector must be set so that the snapshot controller can find PVCs with the matching labels to be snapshotted together.

> ```

Reviewer 2: You can mark this as YAML and get highlighting.

Author: Done

apiVersion: groupsnapshot.storage.k8s.io/v1alpha1
kind: VolumeGroupSnapshot
metadata:
  name: new-group-snapshot-demo
  namespace: demo-namespace
spec:
  volumeGroupSnapshotClassName: csi-groupSnapclass
  source:
    selector:
      group: myGroup
```

In the VolumeGroupSnapshot spec, a user can specify the VolumeGroupSnapshotClass which has the information about which CSI driver should be used for creating the group snapshot.

### Importing an existing group snapshot with Kubernetes
You can always import an existing group snapshot to Kubernetes by manually creating a VolumeGroupSnapshotContent object to represent the existing group snapshot. Because VolumeGroupSnapshotContent is a non-namespace API object, only a system admin may have the permission to create it. Once a VolumeGroupSnapshotContent object is created, the user can create a VolumeGroupSnapshot object pointing to the VolumeGroupSnapshotContent object.

> ```

Reviewer 2: You can mark this as YAML and get highlighting.

Author: Done

apiVersion: groupsnapshot.storage.k8s.io/v1alpha1
kind: VolumeGroupSnapshotContent
metadata:
  name: pre-existing-group-snap-content1
spec:
  driver: com.example.csi-driver
  deletionPolicy: Delete
  source:
     volumeGroupSnapshotHandle: group-snap-id
  volumeGroupSnapshotRef:
    kind: VolumeGroupSnapshot
    name: pre-existing-group-snap1
    namespace: demo-namespace
```

A VolumeGroupSnapshot object should be created to allow a user to use the group snapshot:

> ```

Reviewer 2: You can mark this as YAML and get highlighting.

Author: Done

apiVersion: groupsnapshot.storage.k8s.io/v1alpha1
kind: VolumeGroupSnapshot
metadata:
  name: pre-existing-group-snap1
  namespace: demo-namespace
spec:
  snapshotContentName: pre-existing-group-snap-content1
```

> Once these objects are created, the snapshot controller will bind them together, and set the field Ready (under Status) to True to indicate the group snapshot is ready to use.

Author: Done

* Add group controller capability `CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT`.

See the [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md) and the [Kubernetes-CSI Driver Developer Guide](https://kubernetes-csi.github.io/docs/) for more details.

Although Kubernetes poses as little prescriptive on the packaging and deployment of a CSI Volume Driver as possible, it provides a suggested mechanism to deploy a containerized CSI driver to simplify the process.

As part of this recommended deployment process, the Kubernetes team provides a number of sidecar (helper) containers, including the [external-snapshotter sidecar container](https://kubernetes-csi.github.io/docs/external-snapshotter.html) which has been updated to support volume group snapshot.

The external-snapshotter watches the Kubernetes API server for the `VolumeGroupSnapshotContent` object and triggers `CreateVolumeGroupSnapshot` and `DeleteVolumeGroupSnapshot` operations against a CSI endpoint.

## What are the limitations?
The alpha implementation of volume group snapshots for Kubernetes has the following limitations:

> * Restore is achieved by creating a new PVC from a VolumeSnapshot that is part of a VolumeGroupSnapshot.

Reviewer 2: Why is this a limitation? Perhaps the next item is really part of the same bullet point.

Author: Maybe this should not be listed as a limitation.
In the original design, we are going to introduce a VolumeGroup API object. In that case, we can have an API that allows you to create a VolumeGroup from a VolumeGroupSnapshot in one step. Later on we simplified the design so we no longer have a VolumeGroup construct.
I'll remove this as a limitation.

Author: Done

* Does not support reverting an existing PVC to an earlier state represented by a snapshot that is part of a group snapshot (only supports provisioning a new volume from a snapshot).

> * No application consistency guarantees beyond any guarantees provided by the storage system (e.g. crash consistency).

Reviewer 2: Can you handle that yourself by using, eg, a sidecar to call `fsfreeze`?

Author: Yes

## What’s next?
Depending on feedback and adoption, the Kubernetes team plans to push the CSI Group Snapshot implementation to Beta in either 1.28 or 1.29. Some of the features we are interested in supporting include volume replication, replication group, volume placement, application quiescing, changed block tracking, and more.

## How can I learn more?
The code repository for volume group snapshot APIs and controller is here: https://github.com/kubernetes-csi/external-snapshotter

Check out additional documentation on the group snapshot feature here: https://kubernetes-csi.github.io/docs/

## How do I get involved?

> This project, like all of Kubernetes, is the result of hard work by many contributors from diverse backgrounds working together. We offer a huge thank you to the contributors who stepped up these last few quarters to help the project reach alpha:

Reviewer 2: ```suggestion
This project, like all of Kubernetes, is the result of hard work by many contributors from diverse backgrounds working together. On behalf of SIG Storage, I would like to offer a huge thank you to the contributors who stepped up these last few quarters to help the project reach alpha:
```

“We” does not make sense when there is only one author and without providing more context.

Author: Done
